db/sql_implementation.go

```diff
@@ -0,0 +1,798 @@
+// +build !js
+
+package db
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+
+	"github.com/0xProject/0x-mesh/common/types"
+	"github.com/0xProject/0x-mesh/db/sqltypes"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/gibson042/canonicaljson-go"
+	"github.com/google/uuid"
+	"github.com/ido50/sqlz"
+	"github.com/jmoiron/sqlx"
+	_ "github.com/mattn/go-sqlite3"
+)
+
+var _ Database = (*DB)(nil)
+
+// DB instantiates the DB connection and creates all the collections used by the application
+type DB struct {
+	ctx   context.Context
+	sqldb *sqlz.DB
+	opts  *Options
+}
+
+func defaultOptions() *Options {
+	return &Options{
+		DriverName:     "sqlite3",
+		DataSourceName: "0x_mesh/db/db.sqlite",
+		MaxOrders:      100000,
+		MaxMiniHeaders: 20,
+	}
+}
+
+// TestOptions returns a set of options suitable for testing.
+func TestOptions() *Options {
+	return &Options{
+		DriverName:     "sqlite3",
+		DataSourceName: filepath.Join("/tmp", "mesh_testing", uuid.New().String(), "db.sqlite"),
+		MaxOrders:      100,
+		MaxMiniHeaders: 20,
+	}
+}
+
+// New creates a new connection to the database. The connection will be automatically closed
+// when the given context is canceled.
+func New(ctx context.Context, opts *Options) (*DB, error) {
+	opts = parseOptions(opts)
+
+	connectCtx, cancel := context.WithTimeout(ctx, connectTimeout)
+	defer cancel()
+
+	if err := os.MkdirAll(filepath.Dir(opts.DataSourceName), os.ModePerm); err != nil && err != os.ErrExist {
+		return nil, err
+	}
+
+	sqldb, err := sqlx.ConnectContext(connectCtx, opts.DriverName, opts.DataSourceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Automatically close the database connection when the context is canceled.
```

Comment:
[Both spellings are correct](https://www.grammarly.com/blog/canceled-vs-cancelled/). "canceled" is the American spelling. "cancelled" is the British spelling which I think Fabio favors and it may have rubbed off on me in a few places.

Reply:
> It would be pretty great to have a spell checker for comments because I make typos pretty frequently. It's also extremely low priority, so it's definitely a "nice to have" or even a "wishlist" item.

Feel free to make an issue for this. I have tried various spell checker plugins for VS Code but they have all had issues.

Reply:
Interesting. Thanks for sharing that article. I guess I've always spelled canceled like the Brits (except for this time of course).

I'll open an issue for this.